File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.js
function subarraysDivByK(nums, k) {
    const remainderCount = new Map();
    remainderCount.set(0, 1);

    let prefixSum = 0, count = 0;

    for (let num of nums) {
        prefixSum += num;
        let remainder = ((prefixSum % k) + k) % k; // handle negative mods

        count += remainderCount.get(remainder) || 0;
        remainderCount.set(remainder, (remainderCount.get(remainder) || 0) + 1);
    }

    return count;
}

// Test
const nums = [4, 5, 0, -2, -3, 1];
const k = 5;
console.log(`Number of subarrays divisible by ${k} is: ${subarraysDivByK(nums, k)}`);

File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.go
package main

import (
    "fmt"
)

func subarraysDivByK(nums []int, k int) int {
    remainderCount := make(map[int]int)
    remainderCount[0] = 1

    prefixSum := 0
    count := 0

    for _, num := range nums {
        prefixSum += num
        remainder := ((prefixSum % k) + k) % k // handle negative mod

        count += remainderCount[remainder]
        remainderCount[remainder]++
    }

    return count
}

func main() {
    nums := []int{4, 5, 0, -2, -3, 1}
    k := 5
    result := subarraysDivByK(nums, k)
    fmt.Printf("Number of subarrays divisible by %d is: %d\n", k, result)
}
